src/test/java/uk/gov/companieshouse/web/accounts/util/ValidationHelperTest.java

```diff
@@ -0,0 +1,233 @@
+    private static final Map<String, String> MESSAGE_ARGUMENTS = new HashMap<>();
+
+    @Mock
+    private ClassPathScanningCandidateComponentProvider mockProvider;
+    
+    @InjectMocks
+    private ValidationHelper helper;
+    
+    @BeforeAll
+    public static void setUpData() {
+        MESSAGE_ARGUMENTS.put(MESSAGE_KEY, ERROR);
+    }
+    
+    @Test
+    public void testSuccessfulScan() {
+        mockComponentScanning(BEAN_CLASS_NAME);
+        ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH);
+    }
+    
+    @Test
+    public void testNoMappingsFound() {
+        assertThrows(IllegalStateException.class, () -> ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH));
+    }
+    
+    @Test
+    public void testValidationMappingBeanNotFound() {
+        mockComponentScanning(INVALID_BEAN_CLASS_NAME);
+        assertThrows(IllegalStateException.class, () -> ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH));
+    }
+    
+    @Test
+    public void testNoAnnotationsInValidationModel() {
+        mockComponentScanning(PRIMITIVES_ONLY_BEAN_NAME);
+        assertThrows(IllegalStateException.class, () -> ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH));
+    }
+    
+    @Test
+    public void testScanMaxDepthReached() {
+        mockComponentScanning(RECURSIVE_BEAN_NAME);
+        assertThrows(IllegalStateException.class, () -> ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH));
+    }
+
+    @Test
+    public void testGetValidationError() throws JsonProcessingException {
+        mockComponentScanning(BEAN_CLASS_NAME);
+        ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH);
+        
+        List<ApiError> errors = createErrors(1);
+        ApiErrorResponseException exception = createApiErrorResponse(errors);
+        
+        List<ValidationError> validationErrors = helper.getValidationErrors(exception);
+        assertNotNull(validationErrors);
+        assertEquals(1, validationErrors.size());
+        
+        ValidationError validationError = validationErrors.get(0);
+        assertEquals(FIELD_PATH, validationError.getFieldPath());
+        assertEquals(WEB_ERROR_MESSAGE, validationError.getMessageKey());
+    }
+    
+    @Test
+    public void testGetMultipleValidationErrors() throws JsonProcessingException {
+        mockComponentScanning(BEAN_CLASS_NAME);
+        ValidationHelper.scanPackageForValidationMappings(mockProvider, BASE_PATH);
+        
+        int count = 2;
+        List<ApiError> errors = createErrors(count);
+        ApiErrorResponseException exception = createApiErrorResponse(errors);
+        
+        List<ValidationError> validationErrors = helper.getValidationErrors(exception);
+        assertNotNull(validationErrors);
+        assertEquals(count, validationErrors.size());
+        
+        for(ValidationError validationError : validationErrors) {
```

Comment:
Could use a stream here, not that it's too much more performant for a low number of entries in a set. This is preference so it's entirely up to you

Reply:
It would actually be more performant to do a stream here

Reply:
Based on a 3-execution comparison between a procedural and functional algorithm here (with results averaged and rounded up to two decimal places), the difference in execution time  averaged 0.53 milliseconds in favour of the procedural method used here. Streams aren't a magical performance optimisation, rather they should be used when the context justifies their use.

Reply:
Based on a 3-execution comparison between a procedural and functional algorithm here (with results averaged and rounded up to two decimal places), the difference in execution time  averaged 0.53 milliseconds in favour of the procedural method used here. Streams aren't a magical performance optimisation, rather they should be used when the context justifies their use.

Reply:
parallel streams allow for running the assertions and tests in parallel which will almost certainly be a performance gain, and an improvement going forward as devs learn how to use them. Its as simple as calling parallelStream() or steam().parallel() rather than just stream(). This is not something you can do so easily the procedural way (and you wouldn't want to either).

Reply:
Parallel streams involve a higher overhead than the actual processing effort involved in the loop here. In the case of both `parallelStream()` and `stream().parallel()` both perform worse than the traditional `for` loop.

Given that we're in the context of a unit test here, dealing with a list of fixed-size never likely to change, any performance gain would have negligible benefit, not to mention that cursory testing shows it would actually perform slower. I'm happy to discuss further, but would consider this to be premature optimisation at best.